src/dbnode/storage/namespace/schema.go

```diff
@@ -0,0 +1,169 @@
+	"github.com/golang/protobuf/proto"
+	dpb "github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/jhump/protoreflect/desc"
+	xerrors "github.com/m3db/m3x/errors"
+	"sort"
+	"strings"
+)
+
+var (
+	errInvalidSchema = errors.New("invalid schema")
+)
+
+type schema struct {
+	fdp []byte
+	md  *desc.MessageDescriptor
+	version uint32
+}
+
+func (s *schema) Version() uint32 {
+	return s.version
+}
+
+func (s *schema) Equal(o Schema) bool {
+	return s.Version() == o.Version() && bytes.Equal(s.Bytes(), o.Bytes())
+}
+
+func (s *schema) Get() *desc.MessageDescriptor {
+	return s.md
+}
+
+func (s *schema) String() string {
+	if s.md == nil {
+		return ""
+	}
+	return s.md.String()
+}
+
+func (s *schema) Bytes() []byte {
+	return s.fdp
+}
+
+func schemaListEqual(l, r []Schema) bool {
+	if len(l) != len(r) {
+		return false
+	}
+	for i := 0; i < len(l); i++ {
+		if !l[i].Equal(r[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+func schemaToOption (s Schema) *nsproto.SchemaOption {
+	return &nsproto.SchemaOption{Version: s.Version(), Message: s.Get().GetName(), Definition: s.Bytes()}
+}
+
+func fromSchemaList(schemaList []Schema) []*nsproto.SchemaOption {
+	result := make([]*nsproto.SchemaOption, len(schemaList))
+	for i := 0; i < len(schemaList); i++ {
+		result[i] = schemaToOption(schemaList[i])
+	}
+	return result
+}
+
+// ToSchemaList converts schema options to an array of schema, most recent schema version first.
+func ToSchemaList(options []*nsproto.SchemaOption) ([]Schema, error) {
+	var result []Schema
+	for _, o := range options {
+		s, err := ToSchema(o)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, s)
+	}
+	// sorted by version in decending order
+	sort.Slice(result, func(i, j int) bool {
+		return result[i].Version() > result[j].Version()
+	})
+	return result, nil
+}
+
+func ToSchema(o *nsproto.SchemaOption) (Schema, error) {
+	if o == nil {
+		return nil, nil
+	}
+
+	fdp, err := decodeProtoFileDescriptor(o.Definition)
+	if err != nil {
+		return nil, xerrors.Wrap(err, "failed to decode schema file descriptor")
+	}
+
+	fd, err := desc.CreateFileDescriptor(fdp)
+	if err != nil {
+		return nil, xerrors.Wrap(err, "failed to create schema file descriptor")
+	}
+	mds := fd.GetMessageTypes()
+	if len(mds) == 0 {
+		return nil, xerrors.Wrap(errInvalidSchema, "schema does not contain any messages")
+	}
+
+	var found *desc.MessageDescriptor
+	for _, md := range mds {
+		if strings.EqualFold(md.GetName(), o.Message) {
+			found = md
+		}
+	}
+	if found == nil {
+		return nil, xerrors.Wrapf(errInvalidSchema, "schema does not contain %v", o.Message)
+	}
+
+	return &schema{fdp: o.Definition, md: found, version: o.Version}, nil
+}
+
+// decodeProtoFileDescriptor decodes the bytes of proto file descriptor.
+// proto file descriptor is proto encoded and gzipped, decode reverse the process.
+func decodeProtoFileDescriptor(fdb []byte) (*dpb.FileDescriptorProto, error) {
+	raw, err := decompress(fdb)
```

Comment:
Is the gzipping out of our control? seems like it might be nice to have it human readable even in etcd but maybe its not a big deal

Reply:
This is defined by protobuf. if you look at generated go binding, you will see each file has descriptor bytes embedded in the code already. 

I am not confident on the parsing code, and not quite want to do parsing on our side. I will let the protoc do the parsing and simply have them parse me the equivalent descriptor.

For readability, we can easily rebuild a compact/canonical form of proto by invoking String() method.

Reply:
I think realistically we will want to have them just give us a proto definition and we'll parse it using the dynamic library. Asking people to use protoc is kind of annoying, but we can discuss that when we implement the API